### DISCUSSION TITLE
provider-github long term plan

### DISCUSSION THREAD

**Discussion by jbw976** [crossplane/crossplane#5435] [General]
Multiple folks in the community have brought up recently that the provider-github in the crossplane-contrib org (https://github.com/crossplane-contrib/provider-github) is not actively maintained and hasn't had a commit in a few years.  This came up on the [crossplane-dev](https://groups.google.com/g/crossplane-dev/c/-3YTgGaghNI) mailing list, and again on [Slack #general channel](https://crossplane.slack.com/archives/CEG3T90A1/p1708991989318079).

Meanwhile, there seems to be a good bit of traction around https://github.com/coopnorge/provider-github, which has recent commits and releases.

In general, convergence around providers is a good thing for the community, so this discussion is meant to understand the usage and impact of these two providers and find a path forward that's best for the community.

Some areas to investigate before any decision is made:
* current [`crossplane-contrib/provider-github`](https://github.com/crossplane-contrib/provider-github)
  * Do the current maintainers want to revive this provider and support it long term?
  * How widely used is this provider, would many people be affected by a change to this provider?
  * Is it possible to update the architecture to be based on https://github.com/crossplane/upjet?
* [`coopnorge/provider-github`](https://github.com/coopnorge/provider-github)
  * Does this provider have a long term plan and commitment from its maintainers?
  * Would the maintainers be willing to donate the provider to the crossplane-contrib org?
  * What is the provider architecture/framework, is it based on https://github.com/crossplane/upjet and using best practices to reduce maintenance burden and facilitate ease of adding new resources long term?

We'll follow up with the various parties involved to get a better understanding of the best path forward and share any updates here. In the meantime, opinions and data points are encouraged to be shared here by the community 🙇‍♂️

**Comment by haarchri** [comment]
https://github.com/coopnorge/provider-github/blob/main/Makefile#L9-L14 - is using upjet v1.1.0 https://github.com/coopnorge/provider-github/blob/main/go.mod#L9 as base

**Comment by tomasmota** [comment]
For what it's worth, I have been contributing to and would be willing to help @AtzeDeVries maintain the coopnorge provider.

**Comment by AtzeDeVries** [comment]
I will reply in depth next week, vacationing atm.

**Comment by yordis** [comment]
I apologize if it seems I'm bringing up more questions than providing answers. Additionally, my inquiries extend beyond just the GitHub provider, encompassing a broader scope. Here are my questions:

1. What is the current stance on well-established Terraform providers and their integration with Upjet?
2. Is there a recommendation to deprecate existing non-Upjet repositories in favor of ones prefixed with provider-upjet—*?
3. Could we consider implementing a CI pipeline to streamline much of the code generation process?
4. Is there a detailed and repeatable guide available for migrating to Upjet?
5. IMPORTANT: Has there been a thorough review of CODEOWNERS files across all packages to verify that the Code Owners are active and effectively supporting the project? It's crucial to not only keep the list of names current but also to ensure that maintainers are actively fulfilling their roles. This helps prevent the misconception that repositories are well-managed when, in reality, updates may be necessary.

> What is the provider architecture/framework, is it based on https://github.com/crossplane/upjet and using best practices to reduce maintenance burden and facilitate ease of adding new resources long term?

Extremely important, I have tried multiple times to contribute to and fix DigitalOcean provider, and it has been a harrowing experience. I will try again to see if the situation has improved.

**Comment by AtzeDeVries** [comment]
@jbw976 
> 
> Meanwhile, there seems to be a good bit of traction around https://github.com/coopnorge/provider-github, which has recent commits and releases.
> 
> In general, convergence around providers is a good thing for the community, so this discussion is meant to understand the usage and impact of these two providers and find a path forward that's best for the community.
> 
> Some areas to investigate before any decision is made:
> 
> * [`coopnorge/provider-github`](https://github.com/coopnorge/provider-github)
>   
>   * Does this provider have a long term plan and commitment from its maintainers?

We are using this provider actively and are willing to support this provider. What does long term plan exactly mean? Multiple years? Intentions?  

>   * Would the maintainers be willing to donate the provider to the crossplane-contrib org?
We can consider this. I am not sure what extra is required of us and what we gain etc. 

>   * What is the provider architecture/framework, is it based on https://github.com/crossplane/upjet and using best practices to reduce maintenance burden and facilitate ease of adding new resources long term?

Upjet 1.1.0 is used and we are synced from the template repository in dec/jan 2024. We have not spent a lot of time on E2E tests.

**Reply by jbw976** [reply]
Awesome @AtzeDeVries! yep, really the best we can be asking here is what is your general intentions with your provider - no one can really see multiple years into the future, but understanding that your team is willing to support this provider now is a better place to be in than where https://github.com/crossplane-contrib/provider-github is now 😊 

Thank you for confirming that the https://github.com/coopnorge/provider-github provider is generated with the upjet framework and using recent (v1.0+) patterns. That's a good place to be as well.

So do you have any desire to host this provider in the crossplane-contrib organization and attract community members to collaborate and get involved with more with it? 🙇‍♂️

**Comment by AtzeDeVries** [comment]
I have been in touch with @jbw976  about this. We are planning to move the provider to crossplane-contrib! I will be looking for some maintainers as well. @tomasmota are you still up for it?

**Reply by jbw976** [reply]
Thanks again for the willingness to move this provider to the crossplane-contrib org, that's super cool for the community!

One thing I wanted to personally understand a bit better before we made this move was any impact migration wise. The package manager has long supported a change in package source, e.g. from `xpkg.upbound.io/coopnorge/provider-github:v0.13.0` to `xpkg.upbound.io/crossplane-contrib/provider-upjet-github:v0.13.0`. The package manager should download/install the package from the new location and since the CRDs are all the same there is no underlying migration needed.

However, there was a slight issue in the package manager that existed for awhile and had a workaround here: https://github.com/crossplane/crossplane/issues/3985#issuecomment-1508236672

This issue was fixed in the latest round of patch releases last week, so these versions now have that fix:

* https://github.com/crossplane/crossplane/releases/tag/v1.17.1
* https://github.com/crossplane/crossplane/releases/tag/v1.16.2
* https://github.com/crossplane/crossplane/releases/tag/v1.15.5

After the move, folks on those latest patch versions should be able to just update their `.spec.package` and everything will be happy. If on earlier versions, they can. use the workaround linked above.

We should be able to make this move safely @AtzeDeVries 🙇‍♂️

**Comment by AtzeDeVries** [comment]
So do i understand correctly that people should update their crossplane versions to the once stated above and then the move should be seamless? 
I guess we can add that to the provider documentation and mention it when this move to new organization get published. 

@jbw976 do you think it is wise to wait a bit for people to move to the latest before moving the provider?

**Reply by jbw976** [reply]
> the move should be seamless

That's my understanding of the situation here @AtzeDeVries!

> do you think it is wise to wait a bit for people to move to the latest before moving the provider?

Actually I think we're OK here right now - the old package repo and versions (e.g., `xpkg.upbound.io/coopnorge/provider-github:v0.13.0`) won't be deleted now, so folks on older Crossplane versions can just stay on those until they are ready to move.  When they feel ready to update to say `xpkg.upbound.io/crossplane-contrib/provider-upjet-github:v0.13.0`, then they can also upgrade Crossplane at the same time and everything should be good.

**Comment by jbw976** [comment]
We have a conclusion to share on this discussion 🤩 

https://github.com/coopnorge/provider-github has been successfully transferred to its new home in https://github.com/crossplane-contrib/provider-upjet-github. This provider will start publishing to https://marketplace.upbound.io/providers/crossplane-contrib/provider-upjet-github upon the next official release. Thank you @AtzeDeVries and team for making this generous contribution happen! 

Folks currently using `coopnorge/provider-github` can simply update `.spec.package` of their `Provider` resource to point to the next release version of `xpkg.upbound.io/crossplane-contrib/provider-upjet-github` and they should successfully upgrade without any impact or migration. This is briefly described in https://github.com/crossplane/crossplane/discussions/5435#discussioncomment-10732835.

https://github.com/crossplane-contrib/provider-github has been archived and set to read-only. There will be no development on this provider going forward (as it's been for a couple years now 😂)
